base: Refactor base/random to avoid memory leak

BobbyRBruce opened the pull request:
There was a memory leak occurring prior to this fix which can be reproduced with:

```sh
scons build/VEGA_X86/gem5.opt -j `nproc`

/usr/bin/time -l ./build/VEGA_X86/gem5.opt configs/example/ruby_gpu_random_test.py --WB_L2 --test-length 5000000 --num-dmas 0
```

The `/usr/bin/time -l` will output the maximum memory consumption of this process. On my machines It was getting to 27GB.

With this change gem5 now peaks at around 6GB of memory usage.

This incorporates two changes:

1) Instances of `static RandomPtr` were not creating singular RNG
   instances like assumed. Ergo in casees where this was depended
   upon to stop the over creation of RNGs, multiple were being
   created. The `RandomPtr` object was static but the RNG it pointed
   towards was always changing on every use. This has been fixed.
2) The RNG instance list has been refactored to avoid accidental
   inclusion of the same RNG more than once and better checks have
   been added as to explicitly remove RNGs from this instance list
   when they are deleted.

Tests updated accordingly.

BobbyRBruce committed:
base: Refactor base/random to avoid memory leak

There was a memory leak occuring prior to this fix which can be
reproduced with:

```sh
scons build/VEGA_X86/gem5.opt -j12 `nproc`

/usr/bin/time -l ./build/VEGA_X86/gem5.opt configs/example/ruby_gpu_random_test.py --WB_L2 --test-length 5000000 --num-dmas 0
```

The `/usr/bin/time -l` will output the maximum memory consumption of this
process. On my machines It was getting to 27GB.

With this change gem5 now peaks at around 6GB of memory usage.

This incorporates two changes:

1) Instances of `static RandomPtr` were not creating singular RNG
   instances like assumed. Ergo in casees where this was depended
   upin to stop the over creation of RNGs, multiple were being
   creates. The `RandomPtr` object was static but the RNG it pointed
   towards was always changing on every use.
2) The RNG instance list has been refactored to avoid accidental
   inclusion of the same RNG more than once and better checks have
   been added as to explicitly remove RNGs from this instance list
   when they are deleted.

Tests updated accordingly.

powerjg commented on src/base/random.hh:
```diff
@@ -65,24 +65,26 @@ class Random
+        bool exists = false;
+        for (const auto& weak_ptr : *instances) {
+            if (!weak_ptr.expired() && weak_ptr.lock() == randpoint) {
+            exists = true;
+            break;
```
whitespace

powerjg commented on src/base/random.cc:
```diff
@@ -46,32 +46,30 @@
 namespace gem5
 {
 
+Random::Random(): Random(globalSeed){}
```
Whitespace? I *think* this is a function definition but given that it's all on one line and just above the other definition, I can't tell if that was then intention.

powerjg commented on src/base/random.test.cc:
```diff
@@ -84,7 +84,7 @@ TEST(RandomCtor, DoubleDefaultConstruct)
     // 14514284786278117030llu / std::numeric_limits<uint64_t>::max()
     ASSERT_EQ(dut->random<double>(),
-      0.7868209548678020137657540544751100242137908935546875d);
+      0.7868209548678020137657540544751100242137908935546875);
```
May still need the trailing "d" so that the compiler doesn't truncate.

BobbyRBruce replied:
In CPP literal floating-points are double. The "F" is used to force float and "L" is used to force a long double.

"d" isn't an official CPP convention though GCC permits it. Clang does not which is how I found this bug.

powerjg commented on src/mem/ruby/network/MessageBuffer.cc:
```diff
@@ -205,7 +205,7 @@ MessageBuffer::peek() const
 random_time()
 {
-    static Random::RandomPtr rng = Random::genRandom();
+    Random::RandomPtr rng = Random::genRandom();
```
is this not going to generate a new heap instance of `Random::Random` each time `random_time` is called? We should create one instance and reuse it (e.g., with `static`). At least that is my understanding of the original code.

powerjg commented:
Two things:
1. This wasn't a "leak" (in that it wasn't detected by valgrind --leak-check) because pointers to the random numbers were never out of scope. However, it was a "leak" in the sense that the heap was growing indefinitely when running the `ruby_gpu_random_test`. This was found using the massif tool in `valgrind`.
2. tagging the people involved in the previous refactor: @aperais @odanrc. Tagging @giactra for visibility.

mattsinc reviewed (commented):
As far as I can tell this does not specifically involve a problem with the GPU, but rather VEGA is just being used to demonstrate a specific problem.  So I'll let @powerjg lead the review, but let me know if you need something from me.

odanrc commented on src/base/random.hh:
```diff
@@ -65,24 +65,26 @@ class Random
+        if (instances == nullptr)
+            instances = new Instances();
+
+        auto randpoint = r ? std::shared_ptr<Random>(r, [](Random*){})
```
Is this a call to the constructor with a deleter? I think this line deserves some documentation of why we provide a nop deleter.

powerjg replied:
@BobbyRBruce make sure you address this concern :)

BobbyRBruce replied:
The PR as it currently stands doesn't provide a nop deleter (I did in a previous iteration but I simplified it). However I have added a comment explaining this if-else setting of `randpoint` to alleviate any confusion.

powerjg replied:
What is purpose of the second argument to the constructor of `shared_ptr` here? It looks like an empty deleter, but maybe I'm not understanding correctly.

BobbyRBruce replied:
It's a lambda function which servers as the deletion operator pointer. The purpose here is for the case where the deletion of the pointer is managed elsewhere, which in this case it is as this is.

erin-le committed:
base: Fix compilation failure for base/random refactor

This commit attempts to fix the compilation failure by restoring
an overloaded function that was previously deleted.

powerjg commented on src/base/random.hh:
```diff
@@ -87,6 +87,27 @@ class Random
+        bool exists = false;
+        for (const auto& weak_ptr : *instances) {
+            if (!weak_ptr.expired() && weak_ptr.lock() == randpoint) {
+            exists = true;
+            break;
```
whitespace

powerjg commented:
@BobbyRBruce can we get this in soon?

BobbyRBruce committed:
base: Fix whitespace issues in base/random.cc/.hh

BobbyRBruce committed:
base: Remove new constructor and simplify RNG usage

powerjg commented on src/cpu/testers/gpu_ruby_test/episode.cc:
```diff
@@ -41,6 +41,8 @@
 namespace gem5
 {
 
+//static Random::RandomPtr rng(gem5::Random::genRandom());
```
Probably shouldn't leave comments in the final PR

BobbyRBruce committed:
base: Add comment for genRandom if-else set

BobbyRBruce committed:
cpu: Remove commented out code

BobbyRBruce commented:
I think i addressed all the comments and consider this PR ready for final review and merging. This should fix the Daily Tests.
